src/main/java/net/glowstone/block/blocktype/BlockPiston.java

```diff
@@ -35,17 +46,142 @@ public boolean isSticky() {
+            }
+        }
+
+        // TODO: handle breaking of piston extension
+    }
+
+    @Override
+    public void onRedstoneUpdate(GlowBlock me) {
+        PistonBaseMaterial piston = (PistonBaseMaterial) me.getState().getData();
+        BlockFace pistonBlockFace = piston.getFacing();
+        PistonDirection pistonDirection = PistonDirection.getByBlockFace(pistonBlockFace);
+
+        BlockActionMessage message = new BlockActionMessage(me.getX(), me.getY(), me.getZ(), me.isBlockIndirectlyPowered() ? 0 : 1, pistonDirection.ordinal(), me.getTypeId());
+
+        if (me.isBlockIndirectlyPowered() && !isPistonExtended(me)) {
+            List<Block> blocks = new ArrayList<>();
+
+            // get all blocks to be pushed by piston
+            // add 2 to push limit to compensate for i starting at 1 and also to get the block after the push limit
+            for (int i = 1; i < PUSH_LIMIT + 2; i++) {
+                Block block = me.getRelative(pistonBlockFace, i);
+
+                if (block.getType() == Material.AIR) {
+                    break;
+                }
+
+                // TODO: handle non-pushable blocks.
+
+                // if block after push limit is not air then do not push
+                if (i == PUSH_LIMIT + 1) {
+                    return;
+                }
+
+                blocks.add(block);
+            }
+
+            me.getWorld().getNearbyEntities(me.getLocation(), 20, 20, 20).stream().filter(entity -> entity.getType() == EntityType.PLAYER).forEach(player -> {
+                ((GlowPlayer) player).getSession().send(message);
+            });
+
+            me.getWorld().playSound(me.getLocation(), Sound.BLOCK_PISTON_EXTEND, SoundCategory.BLOCKS, 0.5f, 0.75f);
+
+            // extended state for piston base
+            me.setData((byte) (me.getData() | 0x08));
+
+            for (int i = blocks.size() - 1; i >= 0; i--) {
+                Block block = blocks.get(i);
+
+                setType(block.getRelative(pistonBlockFace), block.getTypeId(), block.getData());
+            }
+
+            // set piston head block when extended
+            setType(me.getRelative(pistonBlockFace), 34, sticky ? me.getData() | 0x08 : pistonDirection.ordinal());
+
             return;
         }
-        if (block.isBlockPowered() || block.isBlockIndirectlyPowered()) {
-            // spawn some smoke because we won't be working on pistons until 1.13
-            block.getWorld().playSound(block.getLocation(), Sound.BLOCK_LAVA_EXTINGUISH, 0.3F, 1.0F);
-            for (int i = 0; i < 3; i++) {
-                block.getWorld().spawnParticle(Particle.SMOKE_LARGE, block.getLocation().clone().add(0.5, 0.8, 0.5), 0);
+
+        if (!isPistonExtended(me)) {
+            return;
+        }
+
+        me.getWorld().getNearbyEntities(me.getLocation(), 20, 20, 20).stream().filter(entity -> entity.getType() == EntityType.PLAYER).forEach(player -> {
```

Comment:
Any reason for the arbitrary `20` blocks range? I think we use this logic usually:

https://github.com/GlowstoneMC/Glowstone/blob/f1d5f6f9fc38c88a414ff2bc60250f7ce151dbd5/src/main/java/net/glowstone/block/entity/BlockEntity.java#L44-L47